crates/extensions/c8y_mapper_ext/src/converter.rs

```diff
@@ -337,6 +343,16 @@ impl CumulocityConverter {
         }
     }
 
+    /// Sanitizes the provided id by replacing any characters not allowed by C8Y
+    /// as an external_id with a safe character like _
+    pub fn sanitize_external_id(id: &str) -> EntityExternalId {
```

Comment:
> We have multiple EntityExternalIds and we don't have a clear way to tell which have these restrictions and which don't

I didn't fully get that statement. Why do you think that we have multiple `EntityExternalId`s with different rules? A single consumer (c8y-mapper in this case) should only have a single kind of external id with a single ruleset applied to it. A different consumer (aws mapper or azure mapper) is free to have a different set of rules. Did you spot the c8y-mapper or any associated component applying different rules to the `EntityExternalId`?

Reply:
I don't mean different rules, but that the consumer has to remember to apply these rules itself, so the "satisfies consumer rules" invariant isn't really represented on the type level. I mean, right now, everything looks correct to me, but otherwise it's easy to forget that one is supposed to use a `sanitize_external_id` function to make sure that we reject invalid ids, instead of just calling `EntityExtrenalId::from`. There's no mistake now, but it will be easy to make a mistake in the future.

External id without any restrictions, and external id with no invalid characters, are currently both a single type. This is becuase `EntityExternalId` is used in the entity store which should support all the clouds, so currently they have to be a single type. But if we make it easier for users of `EntityStore` to attach their own data to entities, we could have `c8y::EntityExternalId`, `az::EntityExternalId`, `aws::EntityExternalId`, etc. all of them guaranteeing that the value inside always meets their own rules and requirements.

Reply:
> but otherwise it's easy to forget that one is supposed to use a sanitize_external_id function to make sure that we reject invalid ids, instead of just calling EntityExtrenalId::from

This is the reason why we are storing these sanitization functions as fields of the `EntityStore`. With that design, the consumer defines those rules in a `Fn` once and passes it on to the entity_store. From then on, the consumer can call `update` with any `id` and the entity store will do that sanitization internally. The external user doesn't have to remember to sanitize before passing anything to the store.

> But if we make it easier for users of EntityStore to attach their own data to entities, we could have c8y::EntityExternalId, az::EntityExternalId, aws::EntityExternalId

Yes, if we could come up with an abstraction like that, that'd be perfect, as that'll help us avoid this function business.